code/parachain/runtime/dali/src/migrations.rs

```diff
@@ -19,6 +21,616 @@ impl OnRuntimeUpgrade for SchedulerMigrationV1toV4 {
+			// check if there is data stored for foreign asset
+			if let Some(foreign_location) = location.clone() {
+				// check that new asset_id is the same as old one for the same location
+				let mut location_stored = false;
+				if let Some(prev_asset_id) =
+					<AssetsRegistry as RemoteAssetRegistryInspect>::location_to_asset(
+						foreign_location.clone(),
+					) {
+					if prev_asset_id != asset_id {
+						panic!("previous and new asset_id for location do not match");
+					}
+					location_stored = true;
+				}
+				// check that new location is the same as old one for the same asset_id
+				let mut asset_stored = false;
+				if let Some(prev_location) =
+					<AssetsRegistry as RemoteAssetRegistryInspect>::asset_to_remote(asset_id)
+				{
+					if prev_location != foreign_location {
+						panic!("previous and new location for asset_id do not match");
+					}
+					asset_stored = true;
+				}
+				// check that either both maps or none map asset_id and location
+				if location_stored != asset_stored {
+					panic!("ForeignToLocal and LocalToForeign maps contradict each other");
+				}
+				if location_stored {
+					<AssetsRegistry as RemoteAssetRegistryMutate>::update_asset(
+						asset_id,
+						asset_info_update(asset_info.clone()),
+					)
+					.expect("asset wasnt updated");
+					count_updated += 1;
+				} else {
+					<AssetsRegistry as RemoteAssetRegistryMutate>::register_asset(
+						asset_id,
+						location.clone(),
+						asset_info.clone(),
+					)
+					.expect("asset wasnt registered");
+					count_created += 1;
+				}
+				continue
+			}
+			// check that for local asset there is no location previously stored
+			if let Some(_prev_location) =
+				<AssetsRegistry as RemoteAssetRegistryInspect>::asset_to_remote(asset_id)
+			{
+				panic!("location is not None for local asset_id");
+			}
+			// check if there is local asset with asset_id
+			if <AssetsRegistry as AssetExistentialDepositInspect>::existential_deposit(asset_id)
+				.is_ok()
+			{
+				<AssetsRegistry as RemoteAssetRegistryMutate>::update_asset(
+					asset_id,
+					asset_info_update(asset_info.clone()),
+				)
+				.expect("asset wasnt updated");
+				count_updated += 1;
+				continue
+			}
+			// register new asset if there was no such asset in memory previously
+			<AssetsRegistry as RemoteAssetRegistryMutate>::register_asset(
+				asset_id, location, asset_info,
+			)
+			.expect("asset wasnt registered");
+			count_created += 1;
+		}
+
+		weights::assets_registry::WeightInfo::<Runtime>::register_asset()
+			.saturating_mul(count_created as u64)
+			.saturating_add(
+				weights::assets_registry::WeightInfo::<Runtime>::update_asset()
+					.saturating_mul(count_updated as u64),
+			)
+	}
+
+	impl OnRuntimeUpgrade for HardCodedAssetsMigration {
+		fn on_runtime_upgrade() -> Weight {
+			let on_chain_version =
+				<AssetsRegistry as GetStorageVersion>::on_chain_storage_version();
+			if on_chain_version < 1 {
```

Comment:
where does the 1 come from?

Reply:
it is the current version we have on dali before upgrade. actually, it is the only place for assets_registry pallet where its version is incremented. the idea is that it would run once. it is common among migrations i looked at

Reply:
hmm can you point to an example?

Reply:
my comment was moreso about the magic number, it should be a constant somewhere

Reply:
https://github.com/paritytech/substrate/pull/9940/files#diff-aef3fe43242285ca7609d95d6fbe9ab1e955d9db07a3de56115fa3b9c40aee6cR34

https://github.com/paritytech/substrate/blob/master/frame/contracts/src/migration.rs#L38